daffodil-cli/build.sbt

```diff
@@ -23,11 +23,21 @@ enablePlugins(JavaAppPackaging)
 enablePlugins(RpmPlugin)
 enablePlugins(WindowsPlugin)
 
+// CLI tests are not thread safe
+// due to use of temporary buffer used to capture stdout/stderr.
+// So we cannot test in parallel
+Test / parallelExecution := false
+IntegrationTest / parallelExecution := false
```

Comment:
The core of the problem is that the the CLI runs from `object Main`, and we mutate the variables STDIN, STDOUT, STDERR in that object everytime we run a CLI test so it uses our `expect` buffers instead of the actual stdin/out/err. Because these are mutated in a single object it's not thread safe.

I guess one thing we might be able to do is change our Main object so it's an actually a class that can be instantiated and is thread safe. Then the object Main just becomes something like this:

```scala
object Main {
  override def main(args: Seq[String]): Unit = {
    val main = new Main()
    main.run(args)
  }
}
```
This would then allow our CLI tests to avoid using the Main object and instead create a new Main instance for each test which can be thread safe, e.g.

```scala
val in, out, err = // get expect buffers
val main = new Main(in, out, err)
main.run(args)
```

I think the one other potential problem area is capturing log4j logs. Like with Main, we also configure log4j to write to these `expect` buffers. But that's also global state and so it's not thread safe either. I don't think there is a way to make log4j have thread local state. Maybe what we can do is have a custom log4j handler that maintains a ThreadLocal of the buffers it should write too.  When we start a thread, we mutate the ThreadLocal with the new buffers to use. And when log4j has a message to write, we get the buffers from the ThreadLocal. I'm not sure if that would actually work, but seems plausible...

Reply:
Note that for the integration tests that fork it is still important to not run in parallel, otherwise we get too much forking and quickly run out of memory.